deeplake/integrations/mmdet/mmdet_.py

```diff
@@ -0,0 +1,309 @@
+from functools import partial
+
+from mmdet.core import BitmapMasks
+import albumentations as A
+import deeplake as dp
+from deeplake.util.warnings import always_warn
+
+
+def _coco_2_pascal(boxes):
+    # Convert bounding boxes to Pascal VOC format and clip bounding boxes to make sure they have non-negative width and height
+    return np.stack(
+        (
+            boxes[:, 0],
+            boxes[:, 1],
+            boxes[:, 0] + np.clip(boxes[:, 2], 1, None),
+            boxes[:, 1] + np.clip(boxes[:, 3], 1, None),
+        ),
+        axis=1,
+    )
+
+
+rand_crop = A.Compose(
+    [
+        A.RandomSizedBBoxSafeCrop(width=128, height=128, erosion_rate=0.2),
+    ],
+    bbox_params=A.BboxParams(
+        format="pascal_voc",
+        label_fields=["labels", "bbox_ids"],
+        min_area=25,
+        min_visibility=0.6,
+    ),
+)
+
+
+def _find_tensor_with_htype(ds: dp.Dataset, htype: str):
+    tensors = [k for k, v in ds.tensors.items() if v.meta.htype == htype]
+    if not tensors:
+        raise ValueError("No tensor found with htype='{htype}'")
+    t = tensors[0]
+    if len(tensors) > 1:
+        always_warn(f"Multiple tensors with htype='{htype}' found. choosing '{t}'.")
+    return t
+
+
+def transform(
+    sample_in,
+    images_tensor: str,
+    masks_tensor: str,
+    boxes_tensor: str,
+    labels_tensor: str,
+    pipeline: Callable,
+):
+    img = sample_in[images_tensor]
+    masks = sample_in[masks_tensor]
+    bboxes = sample_in[boxes_tensor]
+    labels = sample_in[labels_tensor]
+
+    img = img[..., ::-1]
+    if img.shape[2] == 1:
+        img = np.repeat(img, 3, axis=2)
+    masks = masks.transpose((2, 0, 1)).astype(np.uint8)
+    bboxes = _coco_2_pascal(bboxes)
+
+    transformed = rand_crop(
+        image=img,
+        masks=list(masks),
+        bboxes=bboxes,
+        bbox_ids=np.arange(len(bboxes)),
+        labels=labels,
+    )
+
+    img = transformed["image"].astype(np.uint8)
+    shape = img.shape
+    labels = np.array(transformed["labels"], dtype=np.int64)
+    bboxes = np.array(transformed["bboxes"], dtype=np.float32).reshape(-1, 4)
+
+    bbox_ids = transformed["bbox_ids"]
+    masks = [transformed["masks"][i] for i in transformed["bbox_ids"]]
+    if masks:
+        masks = np.stack(masks, axis=0).astype(bool)
+    else:
+        masks = np.empty((0, shape[0], shape[1]), dtype=bool)
+
+    return pipeline(
+        {
+            "img": img,
+            "img_fields": ["img"],
+            "filename": None,
+            "ori_filename": None,
+            "img_shape": shape,
+            "ori_shape": shape,
+            "gt_masks": BitmapMasks(masks, *shape[:2]),
+            "gt_bboxes": bboxes,
+            "gt_labels": labels,
+            "bbox_fields": ["gt_bboxes"],
+        }
+    )
+
+
+def build_dataloader(
+    ds,
+    cfg,
+    images_tensor,
+    masks_tensor,
+    boxes_tensor,
+    labels_tensor,
+    **train_loader_config,
+):
+    if isinstance(ds, dp.Dataset):
```

Comment:
And then here it will become

```suggestion
    if isinstance(ds, dp.Dataset):
        dataloader_builder = DataloaderBuilder(
            ds,
            cfg,
            images_tensor,
            masks_tensor,
            boxes_tensor,
            labels_tensor,
            **train_loader_config,
        )
        return dataloader_builder.build_dataloader()
```